docs/Optimizer.md

```diff
@@ -0,0 +1,48 @@
+# ONNX Optimizer
+
+ONNX provides a C++ library for performing arbitrary optimizations on
+ONNX graphs, as well as a growing list of prepackaged optimization
+passes. The library also provides a convenient in-memory
+representation that is much more convenient to manipulate than the raw
+protobuf structs, and converters to and from the protobuf format.
+
+The primary motivation is to share work between the many ONNX backend
+implementations. Not all possible optimizations can be directly
+implemented on ONNX graphs - some will need additional
+backend-specific information - but many can, and our aim is to provide
+all such passes along with ONNX so that they can be re-used with a
+single function call.
+
+You may be interested in invoking the provided passes, or in
+implementing new ones (or both).
+
+## Invoking The Optimizer
+
+The optimizer may be invoked either via C++ or Python. The Python API
+is described, with example,
+[here](docs/PythonAPIOverview.md#optimizing-an-onnx-model).
+
+The C++ API consists of a single function
+
+```
+ModelProto Optimize(
+    const ModelProto& mp_in,
+    const std::vector<std::string>& names);
+```
+
+which accepts an input `ModelProto` and a list of optimization passes to
+apply, and which returns a new `ModelProto` which is the result of
+applying each of the named passes in sequence to the model. For a list
+of available passes, see [optimize.h](onnx/optimizer/optimize.h).
+
+## Implementing Optimization Passes
+
+You can implement a new optimization pass by subclassing
+`OptimizePass`, and registering your new pass with
+`Optimizer::_registerOptimizer()`. Optimization passes operate on an
+in-memory graph representation defined in [ir.h](onnx/common/ir.h),
```

Comment:
It can also operate on protobuf (and it's sometimes more convenient, e.g. if someone is going to add a pass that populates all the missing optional attributes, it's easier to write it in protobuf, because the default value is stored as AttributeProto in schemas).